src/core/symbology-ng/qgssymbol.cpp

```diff
@@ -172,75 +144,56 @@ QgsConstWkbPtr QgsSymbol::_getLineString( QPolygonF& pts, QgsRenderContext& cont
+QPolygonF QgsSymbol::_getPolygonRing( QgsRenderContext& context, const QgsCurve& curve, bool clipToExtent )
 {
-  QgsWkbTypes::Type wkbType = wkbPtr.readHeader();
-  unsigned int numRings;
-  wkbPtr >> numRings;
-
-  if ( numRings == 0 )  // sanity check for zero rings in polygon
-    return wkbPtr;
-
-  holes.clear();
-
-  QgsCoordinateTransform ct = context.coordinateTransform();
+  const QgsCoordinateTransform ct = context.coordinateTransform();
   const QgsMapToPixel& mtp = context.mapToPixel();
   const QgsRectangle& e = context.extent();
-  double cw = e.width() / 10;
-  double ch = e.height() / 10;
+  const double cw = e.width() / 10;
+  const double ch = e.height() / 10;
   QgsRectangle clipRect( e.xMinimum() - cw, e.yMinimum() - ch, e.xMaximum() + cw, e.yMaximum() + ch );
 
-  int skipZM = ( QgsWkbTypes::coordDimensions( wkbType ) - 2 ) * sizeof( double );
-  Q_ASSERT( skipZM >= 0 );
-
-  for ( unsigned int idx = 0; idx < numRings; idx++ )
-  {
-    unsigned int nPoints;
-    wkbPtr >> nPoints;
+  QPolygonF poly = curve.asQPolygonF();
 
-    if ( static_cast<int>( nPoints * ( 2 * sizeof( double ) + skipZM ) ) > wkbPtr.remaining() )
-    {
-      QgsDebugMsg( QString( "%1 points exceed wkb length (%2>%3)" ).arg( nPoints ).arg( nPoints * ( 2 * sizeof( double ) + skipZM ) ).arg( wkbPtr.remaining() ) );
-      return QgsConstWkbPtr( nullptr, 0 );
-    }
+  if ( curve.numPoints() < 1 )
+    return QPolygonF();
 
-    QPolygonF poly;
-    wkbPtr -= sizeof( unsigned int );
-    wkbPtr >> poly;
-    nPoints = poly.size();
+  //clip close to view extent, if needed
+  const QRectF ptsRect = poly.boundingRect();
+  if ( clipToExtent && !context.extent().contains( ptsRect ) )
+  {
+    QgsClipper::trimPolygon( poly, clipRect );
+  }
 
-    if ( nPoints < 1 )
-      continue;
+  //transform the QPolygonF to screen coordinates
+  if ( ct.isValid() )
+  {
+    ct.transformPolygon( poly );
+  }
 
-    //clip close to view extent, if needed
-    QRectF ptsRect = poly.boundingRect();
-    if ( clipToExtent && !context.extent().contains( ptsRect ) )
-    {
-      QgsClipper::trimPolygon( poly, clipRect );
-    }
+  QPointF *ptr = poly.data();
+  for ( int i = 0; i < poly.size(); ++i, ++ptr )
+  {
+    mtp.transformInPlace( ptr->rx(), ptr->ry() );
+  }
 
-    //transform the QPolygonF to screen coordinates
-    if ( ct.isValid() )
-    {
-      ct.transformPolygon( poly );
-    }
+  return poly;
+}
 
-    QPointF *ptr = poly.data();
-    for ( int i = 0; i < poly.size(); ++i, ++ptr )
-    {
-      mtp.transformInPlace( ptr->rx(), ptr->ry() );
-    }
+void QgsSymbol::_getPolygon( QPolygonF& pts, QList<QPolygonF>& holes, QgsRenderContext &context, const QgsPolygonV2& polygon, bool clipToExtent )
+{
+  holes.clear();
 
-    if ( idx == 0 )
-      pts = poly;
-    else
-      holes.append( poly );
+  //TODO: not working properly
```

Comment:
No. It's something I've fixed and I've forgot to remove the TODO.

Envoyé avec AquaMail pour Android
http://www.aqua-mail.com


Le 24 août 2016 9:15:14 PM Nyall Dawson <notifications@github.com> a écrit :

>>
>> -    if ( idx == 0 )
>> -      pts = poly;
>> -    else
>> -      holes.append( poly );
>> +  //TODO: not working properly
>
> Is this something which should be addressed before merge?
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/qgis/QGIS/pull/3434/files/70defce583ed5522239f5c824236785cf548c746#r76120323